File: Common/Architect.Common/Exceptions/CustomException.vb
Imports System.Globalization
Imports System.Runtime.Serialization
Imports System.Security.Permissions


#Region "Copyright (c) 2007, Global Insurance Technology, Inc."
#End Region

Namespace Exceptions
    <Serializable()>
    Public Class CustomException
        Inherits Exception
        Implements ISerializable

        Private _InvalidFields As Collection

        Protected Sub New(ByVal serializationInfo As SerializationInfo, ByVal streamingContext As StreamingContext)
            MyBase.New(serializationInfo, streamingContext)
        End Sub

        Public Sub New()
            MyBase.New()
        End Sub

        Public Sub New(ByVal message As String)
            MyBase.New(message)
            Helpers.LogHandler.ErrorLog("CustomException", message)
        End Sub

        Public Sub New(ByVal message As String, ByVal inner As Exception)
            MyBase.New(message, inner)
            Helpers.LogHandler.ErrorLog("CustomException", message, inner)
        End Sub

        Public ReadOnly Property InvalidFields() As Collection
            Get
                InvalidFields = _InvalidFields
            End Get
        End Property

        Public Shared Function ShowError(ByVal ex As Exception) As String

            Select Case ex.GetType.Name
                Case "InvalidCastException"
                    Return "Error: Input string was not in a correct format."
                Case "NullReferenceException"
                    Return "Error: Object reference not set to an instance of an object."
                Case Else
                    Return "Unexpected error in the system. Verify with your administrator"
            End Select
        End Function

        <SecurityPermissionAttribute(SecurityAction.LinkDemand, Flags:=SecurityPermissionFlag.SerializationFormatter)>
        Public Overrides Sub GetObjectData(ByVal info As SerializationInfo, ByVal context As StreamingContext) _
            'Implements ISerializable.GetObjectData
            MyBase.GetObjectData(info, context)
            Throw New ArgumentNullException("info")


        End Sub

    End Class

End Namespace

File: Kernel/Architect.Utilities/Exceptions/CustomException.cs
using System;
using System.CodeDom;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Runtime.Serialization;
using System.Security.Permissions;
using System.Text;
using System.Threading.Tasks;

namespace Architect.Utilities.Exceptions
{
    [Serializable()]
    public class CustomException : Exception, ISerializable
    {
        private Collection<Object> _InvalidFields;

        protected CustomException(SerializationInfo serializationInfo, StreamingContext streamingContext) : base(serializationInfo, streamingContext)
        {
        }

        public CustomException() : base()
        {
        }

        public CustomException(string message) : base(message)
        {
            Log.ErrorLog("CustomException", message);
        }

        public CustomException(string message, object data) : base(message)
        {

            Data.Add("Detail", data);
            Log.ErrorLog("CustomException", message);
        }


        public CustomException(string message, Exception inner) : base(message, inner )
        {
            Log.ErrorLog("CustomException", message, inner);
        }

        public CustomException(string message, Exception inner, object data) : base(message, inner)
        {
            Data.Add("Detail", data);
            Log.ErrorLog("CustomException", message, inner);
        }

        public Collection<Object> InvalidFields
        {
            get
            {
                Collection<Object> InvalidFieldsRet = default;
                InvalidFieldsRet = _InvalidFields;
                return InvalidFieldsRet;
            }
        }

        public static string ShowError(Exception ex)
        {
            switch (ex.GetType().Name ?? "")
            {
                case "InvalidCastException":
                    {
                        return "Error: Input string was not in a correct format.";
                    }

                case "NullReferenceException":
                    {
                        return "Error: Object reference not set to an instance of an object.";
                    }

                default:
                    {
                        return "Unexpected error in the system. Verify with your administrator";
                    }
            }
        }

        [SecurityPermission(SecurityAction.LinkDemand, Flags = SecurityPermissionFlag.SerializationFormatter)]
        public override void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            // Implements ISerializable.GetObjectData
            base.GetObjectData(info, context);
            throw new ArgumentNullException("info");
        }

        public string GetStringDataValue(string key)
        {
            string value = string.Empty;

            if (Data.Contains(key))
            {
                value = (string)Data[key];
            }

            return value;
        }
    }
}
